[release/6.0] Fix dump creation on MacOS Ventura

mikem8361 opened the pull request:
Backport of #79356 to release/6.0

Use the task_info(TASK_DYLD_INFO) API to get the dylinker info instead of enumerating all the memory regions. This now works on the new MacOS version (Ventura) and simplifies the module enumeration quite a bit.

Issue: https://github.com/dotnet/runtime/issues/79355

## Customer Impact

The new MacOS version, Ventura, broke core dump generation. This greatly affects VS4Mac Watson triage.

## Testing

The VS4Mac team has verified fix on both x64/arm64 Ventura.  

## Risk

Low risk. Affects only MacOS createdump.

Commit:
Fix dump creation on MacOS Ventura

Use the task_info(TASK_DYLD_INFO) API to get the dylinker info instead of
enumerating all the memory regions. This works on Ventura and simplifies
the module enumeration quite a bit.

jeffschwMSFT reviewed (approved):
approved.  we will take for consideration in 7.0.x

jeffschwMSFT commented:
cc @mmitche

Commit:
Need to ensure the symbol table and symbol string table for the dylinker

module is part of the core dump for the dump readers. They still need
to look up the "dyld_all_image_infos" symbol.

carlossanlop commented:
Approved by Tactics (6.0.14).
Signed off by area owners.
No OOB changes needed.
CI is green.
Ready to merge. :shipit:
